src/sptps_test.c

```diff
@@ -125,6 +134,149 @@ static void usage() {
 	fprintf(stderr, "Report bugs to tinc@tinc-vpn.org.\n");
 }
 
+#ifdef HAVE_MINGW
+
+// Windows does not allow calling select() on anything but sockets. Therefore,
+// to keep the same code as on other operating systems, we have to put a
+// separate thread between the stdin and the sptps loop way below. This thread
+// reads stdin and sends its content to the main thread through a TCP socket,
+// which can be properly select()'ed.
+void *stdin_reader_thread(void *arg) {
+	int *sock_fd = arg;
+	struct sockaddr_in sa;
+	socklen_t sa_size = sizeof(sa);
+
+	while(true) {
+		int peer_fd = accept(*sock_fd, (struct sockaddr *) &sa, &sa_size);
+
+		if(peer_fd < 0) {
+			fprintf(stderr, "accept() failed: %s\n", strerror(errno));
+			continue;
+		}
+
+		if(verbose) {
+			fprintf(stderr, "New connection received from :%d\n", ntohs(sa.sin_port));
+		}
+
+		char buf[1024];
+		ssize_t nread;
+
+		while((nread = read(STDIN_FILENO, buf, sizeof(buf))) > 0) {
+			if(verbose) {
+				fprintf(stderr, "Read %lld bytes from input\n", nread);
+			}
+
+			ssize_t nsend = send(peer_fd, buf, nread, 0);
+
+			if(nsend <= 0) {
+				fprintf(stderr, "Could not send data: %s\n", strerror(errno));
+				break;
+			}
+
+			if(verbose) {
+				fprintf(stderr, "Sent %lld bytes to peer\n", nsend);
+			}
+		}
+
+		shutdown(peer_fd, SD_SEND);
+
+		while(recv(peer_fd, buf, sizeof(buf), 0) > 0) ;
+
+		closesocket(peer_fd);
+	}
+
+	shutdown(*sock_fd, SHUT_RDWR);
+	closesocket(*sock_fd);
+	free(sock_fd);
+}
+
+int start_input_reader() {
+	// Get a random port in range 20000..50000
+	int port = rand() % 30001 + 20000;
+
+	if(verbose) {
+		fprintf(stderr, "Using port %d for the stdin thread\n", port);
+	}
+
+	int sock_fd = socket(AF_INET, SOCK_STREAM, 0);
+
+	if(sock_fd < 0) {
+		fprintf(stderr, "Could not create socket: %s\n", strerror(errno));
+		return -1;
+	}
+
+	struct sockaddr_in addr;
+
+	memset(&addr, 0, sizeof(addr));
+
+	addr.sin_family = AF_INET;
+
+	addr.sin_addr.s_addr = htonl(0x7f000001);
+
+	addr.sin_port = htons(port);
+
+	int res = bind(sock_fd, (struct sockaddr *)&addr, sizeof(addr));
+
+	if(res < 0) {
+		fprintf(stderr, "Could not bind socket: %s\n", strerror(errno));
+		closesocket(sock_fd);
+		return -1;
+	}
+
+	if(listen(sock_fd, 1) < 0) {
+		fprintf(stderr, "Could not listen: %s\n", strerror(errno));
+		closesocket(sock_fd);
+		return -1;
+	}
+
+	int *server_fd = malloc(sizeof(*server_fd));
```

Comment:
That's a bit overkill. I would just store it in a global variable, then there is also no need to pass a pointer to it to  `pthread_create()`.